### PR TITLE
ci: try out the sccache for caching rust code

### PR DESCRIPTION
**Pull request by nagisa**

### CODE REVIEW

**Comment by nagisa** [comment]
cc @Ekleog 

It looks like we’re hitting Github’s rate limits (HTTP 429 code) when interacting with the GH Cache this way :(

This significantly impacts the cache utilization (note: cache write errors)

```
Compile requests                   1262
Compile requests executed           968
Cache hits                          174
Cache hits (C/C++)                   22
Cache hits (Rust)                   152
Cache misses                        793
Cache misses (C/C++)                389
Cache misses (Rust)                 404
Cache timeouts                        0
Cache read errors                     0
Forced recaches                       0
Cache write errors                  792
Compilation failures                  0
Cache errors                          1
Cache errors (Rust)                   1
Non-cacheable compilations            0
Non-cacheable calls                 289
Non-compilation calls                 5
Unsupported compiler calls            0
Average cache write               0.451 s
Average compiler                  1.996 s
Average cache read hit            0.317 s
Failed distributed compilations       0

Non-cacheable reasons:
crate-type                          189
incremental                          70
unknown source language              17
-                                    12
-E                                    1
```

A big benefit to this approach on the other hand is the fact that the cache is shared between different parallel workflows, so if one job gets to and finishes compiling a particular crate first, all of the follow up jobs will be able to immediately re-use the artifact.

**Comment by nagisa** [comment]
So I fiddled with this a bit more, hoping to improve my local setup and I'm finding that even locally `sccache` might not be a win over a fresh build even *with* a hot cache! [Deets for the curious](https://github.com/mozilla/sccache/issues/2041).

**Comment by nagisa** [comment]
nevermind the previous message, sccache does in fact work but gains in my environment do leave a lot to be desired. Definitely should set it up for CI though.

**Comment by Ekleog-NEAR** [comment]
> It looks like we’re hitting Github’s rate limits (HTTP 429 code) when interacting with the GH Cache this way :(

Hmm… is this actually something that’d go away after running 3-4 builds, once the cache is hot? After everything is cached there should be much fewer cache writes (because reads would be hot), so hopefully we could just ignore that and not have to go the full-blown GCS cache way?

**Comment by nagisa** [comment]
15k requests don't sound nearly enough to me (if we do indeed have that as the limit) -- a single fresh `cargo build --workspace` will already trigger ~2k compiler invocations on its own. Then there are like 3 invocations of this entire thing between nextest jobs alone, and we build a `neard` a few times too in some other integration tests.

By my estimation 15k cache lookups would be enough for, like, 2 builds even on a fully primed cache.

**Comment by Ekleog-NEAR** [comment]
Hmm WDYT about just testing it out, given how unclear the github documentation is? I’d suggest we land this, wait for a few days, look at what happened, and rollback if it didn’t actually work out — in this case we can then work on a proper GCS backend

**Comment by nagisa** [comment]
I don’t have a good reason to oppose that.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -192,14 +191,14 @@ jobs:
...
-          save-if: "false" # use the cache from nextest, but don’t double-save
+      - uses: mozilla-actions/sccache-action@1e15924c00ed1446db9e44e8f02628488e7258ec
+      - run: |
+          echo "SCCACHE_GHA_ENABLED=true" >> $GITHUB_ENV
+          echo "RUSTC_WRAPPER=sccache" >> $GITHUB_ENV
       - run: pip3 install --user -r pytest/requirements.txt
       - run: cargo llvm-cov show-env | grep -v RUSTFLAGS | tr -d "'" >> "$GITHUB_ENV"
-      - run: echo "RUSTC_WORKSPACE_WRAPPER=$PWD/scripts/rustc-coverage-wrapper.sh" >> "$GITHUB_ENV"
+      - run: echo "RUSTC_WORKSPACE_WRAPPER=$PWD/scripts/coverage-wrapper-rustc" >> "$GITHUB_ENV"
+      - run: echo "CARGO=1" >> "$GITHUB_ENV"
```
Comment (Ekleog-NEAR) [inline .github/workflows/ci.yml:201]: Just for my understanding, what’s the `CARGO=1` variable about?

**Review by Ekleog-NEAR — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (nagisa) [reply .github/workflows/ci.yml:201]: https://github.com/mozilla/sccache/issues/2040 is the issue to read.

Comment (Ekleog-NEAR) [reply .github/workflows/ci.yml:201]: Thanks!

**Comment by nagisa** [comment]
Looks like we're consistently taking +5 minutes overall after this has landed. Unfortunately the cache failure statistics do not appear to have improved over time, though there are no "read failures" that I can see.

**Pull request merged**